### PR TITLE
Time Suit

### PR DESCRIPTION
**Pull request by Shadowmech88**
image
(Sprites courtesy of @diammz.)
Adds the time suit, a space-worthy lightly-armored suit with time-related capabilities.

Wearing both the suit and the helmet causes the wearer to be timeless, meaning they will be unaffected by external time effects such as chrono grenades.
Additionally, the wearer receives three spells, held inside a unique spellmaster for time spells. These allow the wearer to stop time in a radius of two tiles for five seconds, with a thirty second cooldown, send themselves ten seconds into the future, with a thirty second cooldown, and send themselves ten seconds into the past, with a sixty second cooldown.
WebM of the abilities in combat: http://puu.sh/tTp80/99391f3a95.webm

:cl:
 * rscadd: Added the time suit, a space-worthy lightly-armored suit which confers time-related abilities to the wearer. Sprites by diammz.

### CODE REVIEW

**Comment by Exxion** [comment]
Calling it "sending you to the past" is really kind of a misnomer because the effect is exactly the same as sending you to the future except you can move around and do stuff for the duration rather than just vanishing

**Comment by ghost** [comment]
It resets everything that was in the circle to its original position, health, charge, etc.

**Comment by SonixApache** [comment]
Simple time-suit mobs with one time ability WHEN

**Comment by AlphaBuffalo** [comment]
Might as well be a telelporting suit

**Comment by SkowronX** [comment]
needs a better sprite

**Comment by ghost** [comment]
What's wrong with it?

Hell, what's right with it?

**Comment by PJB3005** [comment]
Shading looks weird.

**Comment by calzilla1** [comment]
Make it when he stops time it has a 1/100 chance to say ZA WARDO

**Comment by Intigracy** [comment]
I loved Timeshift.

Sick reference brah.

**Comment by ghost** [comment]
One step closer to a time cop ERT to remove people who break time law.

**Comment by ghost** [comment]
@calzilla1 That's already a thing with the spell timestop

Not all timestops have to be jojo references

**Comment by Probe1** [comment]
Doesn't the spell time stop play ZA WARDO

**Comment by Exxion** [comment]
Yeah
That's what Skullyton just said

**Comment by Intigracy** [comment]
This specifically is a reference to Timeshift, I knew as soon as I saw the suit sprite.

You're in a time suit that looks just like that and can reverse, freeze, and fast forward time.

**Comment by SonixApache** [comment]
I thought it was Reverse, slow and stop time.

**Comment by kol1th** [comment]
https://github.com/tgstation/tgstation/pull/7084

**Comment by Intigracy** [comment]
Same shit, you moving faster than everything else is you fast forwarding.

**Comment by Intigracy** [comment]
Also who cares about a TG feature tangentially related?

**Comment by NigglyWiggly** [comment]
The shading looks weird on the sprites, yeah.

Not just wearing a DeLorean as a suit/10

**Comment by Shadowmech88** [comment]
Conflicts fixed.

**Comment by kol1th** [comment]
@Intigracy People were talking about time police and that's pretty related.

**Comment by ghost** [comment]
A guy who isn't colourblind did a redo of the sprites

https://my.mixtape.moe/bkwhma.rar

**Comment by diammz** [comment]
i'd be that guy, but i wasn't satisfied with the initial redo so i updated them again.

https://my.mixtape.moe/mfqpoi.rar

**Comment by diammz** [comment]
i'd be that guy, but i wasn't satisfied with the initial redo so i updated it again. (in-hand sprites included)

https://my.mixtape.moe/mfqpoi.rar

suit

**Comment by ghost** [comment]
Nicr

**Comment by diammz** [comment]
custom activation and deactivation sfx for that homey feeling

https://my.mixtape.moe/clzmqu.rar

**Comment by Shadowmech88** [comment]
Better sprites and sounds are added, this should be ready to go.

**Comment by ghost** [comment]
A ready to go PR is kill

**Comment by Shadowmech88** [comment]
I'll get around to fixing it.

**Comment by Shadowmech88** [comment]
Conflicts fixed and ready to go.

File: code/modules/clothing/spacesuits/time.dm [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	armor = list(melee = 25, bullet = 25, laser = 15, energy = 15, bomb = 15, bio = 100, rad = 30)
+	siemens_coefficient = 0.6
+	max_heat_protection_temperature = FIRE_HELMET_MAX_HEAT_PROTECTION_TEMPERATURE
+	clothing_flags = PLASMAGUARD
+	pressure_resistance = 200 * ONE_ATMOSPHERE
+	eyeprot = 3
+
+/obj/item/clothing/head/helmet/space/time/equipped(mob/living/carbon/human/H, head)
+	..()
+	if(istype(H) && H.get_item_by_slot(head) == src)
```
Comment (PJB3005) [inline code/modules/clothing/spacesuits/time.dm]: Should be `slot_head` here.

File: code/modules/clothing/spacesuits/time.dm [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	item_state = "time_helmet"
+	inhand_states = list("left_hand" = 'icons/mob/in-hand/left/clothing.dmi', "right_hand" = 'icons/mob/in-hand/right/clothing.dmi')
+	armor = list(melee = 25, bullet = 25, laser = 15, energy = 15, bomb = 15, bio = 100, rad = 30)
+	siemens_coefficient = 0.6
+	max_heat_protection_temperature = FIRE_HELMET_MAX_HEAT_PROTECTION_TEMPERATURE
+	clothing_flags = PLASMAGUARD
+	pressure_resistance = 200 * ONE_ATMOSPHERE
+	eyeprot = 3
+
+/obj/item/clothing/head/helmet/space/time/equipped(mob/living/carbon/human/H, head)
```
Comment (PJB3005) [inline code/modules/clothing/spacesuits/time.dm]: Why is the second parameter called `head` anyways?

File: code/modules/clothing/spacesuits/time.dm [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	suit_active = FALSE
+	H.flags &= ~TIMELESS
+	H.remove_spell(timestop)
+	H.remove_spell(futurejump)
+	H.remove_spell(pastjump)
+	playsound(src, 'sound/misc/timesuit_deactivate.ogg', 50)
+
+/obj/item/clothing/suit/space/time/equipped(mob/living/carbon/human/H, wear_suit)
+	..()
+	if(istype(H) && H.get_item_by_slot(wear_suit) == src)
```
Comment (PJB3005) [inline code/modules/clothing/spacesuits/time.dm]: `slot_wear_suit`

File: code/modules/clothing/spacesuits/time.dm [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	if(!istype(H))
+		return
+	suit_active = FALSE
+	H.flags &= ~TIMELESS
+	H.remove_spell(timestop)
+	H.remove_spell(futurejump)
+	H.remove_spell(pastjump)
+	playsound(src, 'sound/misc/timesuit_deactivate.ogg', 50)
+
+/obj/item/clothing/suit/space/time/equipped(mob/living/carbon/human/H, wear_suit)
```
Comment (PJB3005) [inline code/modules/clothing/spacesuits/time.dm]: Naming of the second argument.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: code/game/objects/items/weapons/grenades/chronogrenade.dm [hunk 5]
```diff
@@ -37,14 +37,23 @@
...
-	for(var/turf/T in targets)
-		for(var/atom/movable/everything in T)
-			if(everything.flags & TIMELESS || everything.being_sent_to_past)	//allowing future grenades to interact with past-tethered atoms would be a nightmare
-				continue
-			everything.send_to_future(duration)
-			if(ismob(everything))
-				var/mob/M = everything
-				M.playsound_local(everything, 'sound/effects/fall2.ogg', 100, 0, 0, 0, 0)
+	if(single_target)
+		if(!A.being_sent_to_past)
```
Comment (ComicIronic) [inline code/game/objects/items/weapons/grenades/chronogrenade.dm]: So a single target can be affected even if timeless? Also, there's a lot of copypaste to all of this - make these procs.

**Comment by Shadowmech88** [comment]
*Now* it should be ready.

**Comment by ghost** [comment]
Where do you get this? Bus only?

**Comment by Shadowmech88** [comment]
For now.

**Review by PJB3005 — APPROVED** [review]

**Comment by ComicIronic** [comment]
Approved, merging.

**Pull request merged**